typing/typecore.ml

```diff
@@ -2741,14 +2762,18 @@ let collect_unknown_apply_args env funct ty_fun mode_fun rev_args sargs ret_tvar
                 Location.prerr_warning sarg.pexp_loc
                   Warnings.Ignored_extra_argument;
               let mode_arg = Alloc_mode.newvar () in
-              let mode_res = Alloc_mode.newvar () in
-              let kind = (lbl, mode_arg, mode_res) in
+              let mode_ret = Alloc_mode.newvar () in
+              check_local_application_complete ~env ~app_loc ~sarg ~rest
+                ~mode_fun ~mode_arg ~mode_ret;
+              let kind = (lbl, mode_arg, mode_ret) in
               unify env ty_fun
                 (newty (Tarrow(kind,ty_arg,ty_res,commu_var ())));
-              (mode_arg, ty_arg_mono, mode_res, ty_res)
-        | Tarrow ((l, mode_arg, mode_res), ty_arg, ty_res, _)
+              (mode_arg, ty_arg_mono, mode_ret, ty_res)
+        | Tarrow ((l, mode_arg, mode_ret), ty_arg, ty_res, _)
           when labels_match ~param:l ~arg:lbl ->
-            (mode_arg, tpoly_get_mono ty_arg, mode_res, ty_res)
+            check_local_application_complete ~env ~app_loc ~sarg ~rest
```

Comment:
Why don't the checks here need to be delayed like elsewhere?

Reply:
In this case, we're looking at an application to an unknown function and inferring the function's type, and it wasn't obvious to me that the default order of checks was wrong. (In the other case, we *know* the type of the function, and we should really use that known information before we start messing about with modes).

(Having said that, I've no problem with delaying the checks here too. I don't feel strongly either way so I did the thing that took the least code)

Reply:
Make sense to me.